File: algorithm_practice_python/leetcode/default/best_time_buy_sell_stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # prices[i] = i번째날 가격
        # 가장 싸게 사서 비싸게 팔때 최대 이윤
        # 현재 위치 - 나머지 위치(n) = max를 찾아라

        max_profit = 0
        min_price = float('inf')
        for price in prices:
            min_price = min(price, min_price)
            profit = price - min_price
            max_profit = max(profit, max_profit)

        return max_profit

File: algorithm_practice_java/src/main/java/leetcode/default/BestTimeBuySellStock.kt
package leetcode.default

import kotlin.math.max

class Solution {
    //    fun maxProfit(prices: IntArray): Int {
//        var maxProfit = 0
//        var minPrice = Int.MAX_VALUE
//        for (price in prices) {
//            minPrice = min(price, minPrice)
//            maxProfit = max(price - minPrice, maxProfit)
//        }
//        return maxProfit
//    }
    //using kadane's algorithm
    // maxcur += max(0, price[i]- price[i-1])
    fun maxProfit(prices: IntArray): Int {
        var maxCur = 0
        var maxSoFar = 0

        (1 until prices.size).forEach {
            maxCur += prices[it] - prices[it - 1]
            maxCur = max(0, maxCur)
            maxSoFar = max(maxCur, maxSoFar)
        }

        return maxSoFar
    }
}